exonum/src/explorer/mod.rs

```diff
@@ -0,0 +1,976 @@
+///
+/// - `type` field contains transaction type (`"in-pool"`).
+/// - `content` is JSON serialization of the transaction.
+///
+/// ```
+/// # #[macro_use] extern crate serde_json;
+/// # extern crate exonum;
+/// # #[cfg(feature = "doctests")] fn main() {
+/// # use exonum::explorer::{BlockchainExplorer, TransactionInfo};
+/// # use exonum::explorer::tests::{sample_blockchain, mempool_transaction};
+/// use exonum::encoding::serialize::json::ExonumJson;
+///
+/// # let blockchain = sample_blockchain();
+/// # let explorer = BlockchainExplorer::new(&blockchain);
+/// let tx_in_pool: TransactionInfo = // ...
+/// #   explorer.transaction(&mempool_transaction().hash()).unwrap();
+/// assert_eq!(
+///     serde_json::to_value(&tx_in_pool).unwrap(),
+///     json!({
+///         "type": "in-pool",
+///         "content": tx_in_pool.content().serialize_field().unwrap(),
+///     })
+/// );
+/// # } // main
+/// # #[cfg(not(feature = "doctests"))] fn main() {}
+/// ```
+#[derive(Debug, Serialize)]
+#[serde(tag = "type", rename_all = "kebab-case")]
+pub enum TransactionInfo {
+    /// Transaction is in the memory pool, but not yet committed to the blockchain.
+    InPool {
+        /// Transaction contents.
+        #[serde(serialize_with = "CommittedTransaction::serialize_content")]
+        content: Box<Transaction>,
+    },
+
+    /// Transaction is already committed to the blockchain.
+    Committed(CommittedTransaction),
+}
+
+impl TransactionInfo {
+    /// Returns the content of this transaction.
+    pub fn content(&self) -> &Transaction {
+        match *self {
+            TransactionInfo::InPool { ref content } => content.as_ref(),
+            TransactionInfo::Committed(ref tx) => tx.content(),
+        }
+    }
+
+    /// Is this in-pool transaction?
+    pub fn is_in_pool(&self) -> bool {
+        match *self {
+            TransactionInfo::InPool { .. } => true,
+            _ => false,
+        }
+    }
+
+    /// Is this a committed transaction?
+    pub fn is_committed(&self) -> bool {
+        match *self {
+            TransactionInfo::Committed(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns a reference to the inner committed transaction if this transaction is committed.
+    /// For transactions in pool, returns `None`.
+    pub fn as_committed(&self) -> Option<&CommittedTransaction> {
+        match *self {
+            TransactionInfo::Committed(ref tx) => Some(tx),
+            _ => None,
+        }
+    }
+}
+
+/// Information on blocks coupled with the corresponding range in the blockchain.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BlocksRange {
+    /// Exclusive range of blocks.
+    pub range: Range<Height>,
+    /// Blocks in the range.
+    pub blocks: Vec<Block>,
+}
+
+/// Blockchain explorer.
+pub struct BlockchainExplorer<'a> {
+    snapshot: Box<Snapshot>,
+    transaction_parser: Box<'a + Fn(RawMessage) -> ParseResult>,
```

Comment:
The external lifetime is relatively easy to get rid of (by cloning the `Blockchain` in the constructor), but imo it serves a purpose: it makes it impossible to update the blockchain while there is an instance of explorer for it; i.e., it makes use of Rust borrow checker.